### PR TITLE
Allow use of GSM for communal credential secret

### PR DESCRIPTION
**Pull request by sivaalamp**
This adds a feature flag to get the secret for the communal credential from google secret manager (GSM) instead of from a k8s secret. 

It is off by default. You must set the following annotation in the VerticaDB CR:
```
kind: VerticaDB
metadata:
  name: v
  annotations:
    vertica.com/use-gcp-secret-manager: "true" 
spec:
  communal:
    credentialSecret: <full name of the secret in GSM>
...
```

The name of the GSM secret is taken from the existing `spec.communal.credentialSecret` field.

### CODE REVIEW

File: pkg/controllers/vdb/init_db.go [hunk 1]
```diff
@@ -269,12 +279,55 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+	result, err := client.AccessSecretVersion(ctx, req)
+	if err != nil {
+		g.Log.Info("google: could not find default credentials")
+		return "", ctrl.Result{}, err
+	}
+
+	crc32c := crc32.MakeTable(crc32.Castagnoli)
+	checksum := int64(crc32.Checksum(result.Payload.Data, crc32c))
+	if checksum != *result.Payload.DataCrc32C {
+		return "", ctrl.Result{}, err
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:303]: err would be nil here. We should generate one:
```suggestion
		return "", ctrl.Result{}, fmt.Errorf("checksum failure, expected %d but got %d", *result.Payload.DataCrc32C, checksum)```

File: pkg/controllers/vdb/init_db.go [hunk 2]
```diff
@@ -269,12 +279,55 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+	}
+
+	var gs GcpSecret
+	err = json.Unmarshal([]byte(string(result.Payload.Data)), &gs)
+	if err != nil {
+		g.Log.Info("Permission 'secretmanager.versions.access' denied for resource to read the secret")
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:309]: We should use `g.Log.Error`

Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:309]: Why do we consider this a permission issue? Couldn't we get this error if the JSON was in a different format (i.e. key values don't match or additional fields are included).

File: pkg/controllers/vdb/init_db.go [hunk 3]
```diff
@@ -269,12 +279,55 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+	}
+	defer client.Close()
+
+	req := &secretmanagerpb.AccessSecretVersionRequest{
+		Name: name,
+	}
+
+	result, err := client.AccessSecretVersion(ctx, req)
+	if err != nil {
+		g.Log.Info("google: could not find default credentials")
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:296]: We should use `g.Log.Error`

Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:296]: What does the error message mean? What are default credentials

File: pkg/controllers/vdb/init_db.go [hunk 4]
```diff
@@ -65,6 +70,11 @@ type GenericDatabaseInitializer struct {
 	ConfigurationParams *vtypes.CiMap
 }
 
+type GcpSecret struct {
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:73]: Lets not export this struct. Also, secret is too generic. Lets call out the fact that it's for credentials.
```suggestion
type gcpCredentialSecret struct {
```

File: pkg/controllers/vdb/init_db.go [hunk 5]
```diff
@@ -65,6 +70,11 @@ type GenericDatabaseInitializer struct {
...
+	Accesskey string
+	Secretkey string
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:75]: Does this need to have specific field names, or could we have used something more generic like `map[string]string` instead?

File: pkg/controllers/vdb/init_db.go [hunk 6]
```diff
@@ -65,6 +70,11 @@ type GenericDatabaseInitializer struct {
 	ConfigurationParams *vtypes.CiMap
 }
 
+type GcpSecret struct {
+	Accesskey string
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:74]: Depending on other comments, this may not apply anymore. But I'm including it here so I don't forget... since we use struct to unmarshal JSON, we should have JSON style comments attached with it.
```suggestion
	Accesskey string ``json:"accesskey"`
```

File: pkg/controllers/vdb/init_db.go [hunk 7]
```diff
@@ -269,12 +279,55 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+	}
+
+	crc32c := crc32.MakeTable(crc32.Castagnoli)
+	checksum := int64(crc32.Checksum(result.Payload.Data, crc32c))
+	if checksum != *result.Payload.DataCrc32C {
+		return "", ctrl.Result{}, err
+	}
+
+	var gs GcpSecret
+	err = json.Unmarshal([]byte(string(result.Payload.Data)), &gs)
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:307]: What data type is `result.Payload.Data`?

File: pkg/controllers/vdb/init_db.go (hunk 3 above)
Comment (roypaulin) [inline pkg/controllers/vdb/init_db.go:296]: Where is it supposed to find the credentials? By credentials do you mean credentials for the GCP account?

File: pkg/meta/annotations.go [hunk 8]
```diff
@@ -37,6 +37,12 @@ const (
 	// treated as a boolean.
 	VClusterOpsAnnotation     = "vertica.com/vcluster-ops"
 	VClusterOpsAnnotationTrue = "true"
+
+	// This is a feature flag for accessing the secrets configured in Google Secret Manager.
+	// This is set to true by default to let the verticadb running cluster to access the secrets.
+	// The value of this annotation is treated as a boolean.
+	GcpGsmAnnotation     = "vertica.com/gcpgsmaccess"
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/meta/annotations.go:44]: Changed the annotation

File: pkg/meta/annotations.go [hunk 9]
```diff
@@ -51,6 +57,12 @@ func UseVClusterOps(annotations map[string]string) bool {
 	return lookupBoolAnnotation(annotations, VClusterOpsAnnotation, false)
 }
 
+// UseGCPSecretManager returns true nk all admin commands should use the vclusterOps
+// library rather than admintools.
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/meta/annotations.go:61]: Added the appropriate comments

File: pkg/meta/annotations.go [hunk 10]
```diff
@@ -37,6 +37,12 @@ const (
 	// treated as a boolean.
 	VClusterOpsAnnotation     = "vertica.com/vcluster-ops"
 	VClusterOpsAnnotationTrue = "true"
+
+	// This is a feature flag for accessing the secrets configured in Google Secret Manager.
+	// This is set to true by default to let the verticadb running cluster to access the secrets.
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/meta/annotations.go:42]: The default value is changed to false now and removed the comment about the same

File: pkg/controllers/vdb/init_db.go [hunk 11]
```diff
@@ -17,16 +17,21 @@ package vdb
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"hash/crc32"
 	"regexp"
 	"strings"
 
+	gsm "cloud.google.com/go/secretmanager/apiv1"
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:26]: will check in the go.mod and go.sum files

File: pkg/controllers/vdb/init_db.go [hunk 12]
```diff
@@ -269,12 +279,55 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
 	g.ConfigurationParams.Set("KerberosEnableKeytabPermissionCheck", "0")
 }
 
+func (g *GenericDatabaseInitializer) setAuthFromGCSSecret(ctx context.Context) (string, ctrl.Result, error) {
+	name := "projects/otl-ot2-int-sb/secrets/new_vertica_secret/versions/1"
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:283]: Introduced the annotations to read the project id and secret name

File: pkg/controllers/vdb/init_db.go (hunk 1 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:303]: Added err as suggested

File: pkg/controllers/vdb/init_db.go (hunk 2 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:309]: Changed to Error instead of Info

File: pkg/controllers/vdb/init_db.go (hunk 3 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:296]: Changed to Error instead of Info

File: pkg/controllers/vdb/init_db.go (hunk 4 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:73]: This comment is obsolete now as I changed it to map[string]string{} and this struct is no more present

File: pkg/controllers/vdb/init_db.go (hunk 5 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:75]: Removed the struct and introduced map

File: pkg/controllers/vdb/init_db.go (hunk 6 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:74]: This comment is obsolete now as I changed it to map[string]string{} and this struct is no more present

File: pkg/controllers/vdb/init_db.go (hunk 7 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:307]: it is an uint8[]

File: pkg/controllers/vdb/init_db.go (hunk 3 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go]: Updated with appropriate error message

Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go]: Yes.

File: pkg/controllers/vdb/init_db.go [hunk 13]
```diff
@@ -269,12 +274,58 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
 	g.ConfigurationParams.Set("KerberosEnableKeytabPermissionCheck", "0")
 }
 
+func (g *GenericDatabaseInitializer) setAuthFromGCSSecret(ctx context.Context) (string, ctrl.Result, error) {
+	name := "projects/" + g.Vdb.Annotations[meta.GcpProjectIDAnnotation] +
+		"/secrets/" + g.Vdb.Annotations[meta.GcpSecNameAnnontation] +
+		"/versions/" + g.Vdb.Annotations[meta.GcpSecVersionAnnotation]
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:280]: We shouldn't use annotations to come up with then secret name. We already have a field in the VerticaDB CR for the name and should use that: `spec.communal.credentialSecret`

File: pkg/controllers/vdb/init_db.go (hunk 1 above)
Comment (spilchen) [reply pkg/controllers/vdb/init_db.go:302]: err is still nil here I believe.

File: pkg/controllers/vdb/init_db.go [hunk 14]
```diff
@@ -269,12 +274,58 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
 	g.ConfigurationParams.Set("KerberosEnableKeytabPermissionCheck", "0")
 }
 
+func (g *GenericDatabaseInitializer) setAuthFromGCSSecret(ctx context.Context) (string, ctrl.Result, error) {
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:277]: Let's drop `ctrl.Result` as an output parm. I think we set it to `ctrl.Result{}` each time.

File: pkg/controllers/vdb/init_db.go (hunk 3 above)
Comment (spilchen) [reply pkg/controllers/vdb/init_db.go]: It is my understanding that the host where the operator runs has a profile setup to allow it access to the GSM.

File: pkg/controllers/vdb/init_db.go (hunk 7 above)
Comment (spilchen) [reply pkg/controllers/vdb/init_db.go]: I think uint8 is the same as byte. Can we just pass in the payload without a cast?

File: pkg/controllers/vdb/init_db.go (hunk 14 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:277]: Making use of it now, so didn't change the return type

File: pkg/controllers/vdb/init_db.go (hunk 13 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:280]: Removed the new annotations and used the existing credentialSecret for GSM as well

File: pkg/controllers/vdb/init_db.go (hunk 1 above)
Comment (spilchen) [reply pkg/controllers/vdb/init_db.go:299]: I know we talked about this on teams, but the code here looks wrong. I think my suggestion will solve it.

File: pkg/controllers/vdb/init_db.go [hunk 15]
```diff
@@ -269,12 +274,70 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+	name := g.Vdb.Spec.Communal.CredentialSecret
+	client, err := gsm.NewClient(ctx)
+	if err != nil {
+		return "", ctrl.Result{}, err
+	}
+	defer client.Close()
+
+	req := &secretmanagerpb.AccessSecretVersionRequest{
+		Name: name,
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:286]: Suggest removal of the `name` parameter:
```suggestion
		Name: g.Vdb.Spec.Communal.CredentialSecret,
```

File: config/samples/v1beta1_verticadb.yaml [hunk 16]
```diff
@@ -15,12 +15,13 @@ apiVersion: vertica.com/v1beta1
 kind: VerticaDB
 metadata:
   name: verticadb-sample
+
```
Comment (spilchen) [inline config/samples/v1beta1_verticadb.yaml:18]: minor nit: can we just undo the changes in here entirely. I know semantically nothing changed, I would prefer if this file didn't change at all

File: pkg/meta/annotations.go [hunk 17]
```diff
@@ -37,6 +37,11 @@ const (
 	// treated as a boolean.
 	VClusterOpsAnnotation     = "vertica.com/vcluster-ops"
 	VClusterOpsAnnotationTrue = "true"
+
+	// This is a feature flag for accessing the secrets configured in Google Secret Manager.
+	// The value of this annotation is treated as a boolean.
+	GcpGsmAnnotation     = "vertica.com/use-gcp-secret-manager"
+	GcpGsmAnnotationTrue = "true"
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/meta/annotations.go:44]: These should be supplied in the yaml file to return true from UseGCPSecretManager so that it fetches gsm secrets.

File: pkg/controllers/vdb/init_db.go (hunk 15 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:286]: removed the name variable

File: pkg/controllers/vdb/init_db.go (hunk 1 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:299]: Was a copy paster error, corrected it now

File: pkg/controllers/vdb/init_db.go (hunk 7 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go]: No casting is done

File: pkg/controllers/vdb/init_db.go (hunk 3 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go]: Added the proper logs now

File: pkg/controllers/vdb/init_db.go [hunk 18]
```diff
@@ -269,12 +274,70 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+			"The communal credential secret '%s' does not have a key named '%s'", g.Vdb.Spec.Communal.CredentialSecret, cloud.CommunalAccessKeyName)
+		return "", ctrl.Result{Requeue: true}, nil
+	}
+
+	secretKey, ok := GcpCred[cloud.CommunalSecretKeyName]
+	if !ok {
+		g.VRec.Eventf(g.Vdb, corev1.EventTypeWarning, events.CommunalCredsWrongKey,
+			"The communal credential secret '%s' does not have a key named '%s'", g.Vdb.Spec.Communal.CredentialSecret, cloud.CommunalSecretKeyName)
+		return "", ctrl.Result{Requeue: true}, nil
+	}
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:321]: Not getting the auth value in proper readable format when the GcpCred is changed to map[string][]byte{}, so the code is retained

File: pkg/controllers/vdb/init_db.go [hunk 19]
```diff
@@ -269,12 +274,70 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+	if checksum != *result.Payload.DataCrc32C {
+		g.Log.Error(err, "Permission denied to read the secret")
+		return "", ctrl.Result{}, err
+	}
+
+	GcpCred := map[string]string{}
+
+	err = json.Unmarshal([]byte(string(result.Payload.Data)), &GcpCred)
+	if err != nil {
+		return "", ctrl.Result{}, fmt.Errorf("checksum failure, expected %d but got %d", *result.Payload.DataCrc32C, checksum)
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:306]: The error text does make sense here. Is this a copy+paste error?

File: pkg/controllers/vdb/init_db.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (spilchen) [reply pkg/controllers/vdb/init_db.go:321]: Can you provide more details here? Is one of the values base64 encoded?

File: pkg/controllers/vdb/init_db.go [hunk 20]
```diff
@@ -269,12 +274,70 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+	}
+
+	secretKey, ok := GcpCred[cloud.CommunalSecretKeyName]
+	if !ok {
+		g.VRec.Eventf(g.Vdb, corev1.EventTypeWarning, events.CommunalCredsWrongKey,
+			"The communal credential secret '%s' does not have a key named '%s'", g.Vdb.Spec.Communal.CredentialSecret, cloud.CommunalSecretKeyName)
+		return "", ctrl.Result{Requeue: true}, nil
+	}
+
+	auth := fmt.Sprintf("%s:%s", strings.TrimSuffix(string(accessKey), "\n"), strings.TrimSuffix(string(secretKey), "\n"))
```
Comment (spilchen) [inline pkg/controllers/vdb/init_db.go:323]: Why do you need to strip newlines from the secrets?

Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:323]: the configuration parameter takes the auth value in terms of accesskey:secretkey, which forming the auth parameter to format in the required format if there are any new lines those needs to be trimmed. I already have checked, if there are new lines in the keys, nothing will happen to the string.

File: pkg/controllers/vdb/init_db.go (hunk 19 above)
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go:306]: Moved it already

File: pkg/meta/annotations.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (spilchen) [reply pkg/meta/annotations.go:44]: But do we need this const defined in the code here? Where do we use it?

File: pkg/controllers/vdb/init_db.go (hunk 20 above)
Comment (spilchen) [reply pkg/controllers/vdb/init_db.go]: But did you see a case where a newline was at the end of one of the keys?

**Comment by spilchen** [comment]
@sivaalamp we are almost finished with the review. There are just two remaining comments to answer and/or address.

File: pkg/controllers/vdb/init_db.go [hunk 21]
```diff
@@ -269,12 +274,70 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+
 // setGCloudAuthParms adds the auth parms to the config parms map when we are
 // connecting to google cloud storage.
 func (g *GenericDatabaseInitializer) setGCloudAuthParms(ctx context.Context) (ctrl.Result, error) {
-	res, err := g.setAuth(ctx, "GCSAuth")
-	if verrors.IsReconcileAborted(res, err) {
-		return res, err
+	if meta.UseGCPSecretManager(g.Vdb.Annotations) {
+		auth, res, err := g.setAuthFromGCSSecret(ctx)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/controllers/vdb/init_db.go]: Added as suggested

File: pkg/meta/annotations.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (sivaalamp) [reply pkg/meta/annotations.go:44]: Removed the GcpGsmAnnotationTrue constant, as we can directly supply value true from the yaml file

File: config/samples/v1beta1_verticadb.yaml [hunk 22]
```diff
@@ -15,10 +15,16 @@ apiVersion: vertica.com/v1beta1
 kind: VerticaDB
 metadata:
   name: verticadb-sample
+  annotations:
```
[the remark this answers is not in the document]
Comment (sivaalamp) [reply config/samples/v1beta1_verticadb.yaml]: no change in this file now

File: pkg/controllers/vdb/init_db.go [hunk 23]
```diff
@@ -269,12 +274,68 @@ func (g *GenericDatabaseInitializer) setKerberosAuthParms() {
...
+		return "", ctrl.Result{}, fmt.Errorf("data corruption detected, expected %d but got %d", *result.Payload.DataCrc32C, checksum)
+	}
+
+	GcpCred := map[string]string{}
+	err = json.Unmarshal(result.Payload.Data, &GcpCred)
+	if err != nil {
+		return "", ctrl.Result{}, err
+	}
+
+	accessKey, ok := GcpCred[cloud.CommunalAccessKeyName]
```
[the remark this answers is not in the document]
Comment (roypaulin) [reply pkg/controllers/vdb/init_db.go:307]: I am fine with adding a comment. I couldn't find an easy way to remove the duplication.

**Review by spilchen — APPROVED** [review]
